--- bash-4.1/subst.c	2015-04-07 10:00:00.482931289 +0900
+++ bash-4.1/subst.c	2015-04-07 10:01:28.258111134 +0900
@@ -7039,8 +7039,6 @@ 
 
       ret = alloc_word_desc ();
       ret->word = temp1;
-      ret = alloc_word_desc ();
-      ret->word = temp1;
       if (temp1 && QUOTED_NULL (temp1) && (quoted & (Q_HERE_DOCUMENT|Q_DOUBLE_QUOTES)))
 	ret->flags |= W_QUOTED|W_HASQUOTEDNULL;
       return ret;
